libs/s25main/ai/aijh/AIPlayerJH.cpp

```diff
@@ -1442,11 +1457,14 @@ void AIPlayerJH::MilUpgradeOptim()
             {
                 aii.SetCoinsAllowed(milBld->GetPos(), true);
             }
-            if(milBld->HasMaxRankSoldier()) // has max rank soldier? send it/them out!
-                aii.SendSoldiersHome(milBld->GetPos());
-            if(SoldierAvailable(0)
-               && milBld->GetNumTroops() < milBld->GetMaxTroopsCt()) // building not full and privates in a warehouse?
-                aii.OrderNewSoldiers(milBld->GetPos());              // order new!
+            // Keep 0 generals, 1 of each other rank and fill the rest with privates
+            unsigned rank = 0;
+            for(unsigned count : {milBld->GetMaxTroopsCt(), 1u, 1u, 1u, 0u})
```

Comment:
I think you should use an array of size `NUM_SOLDIER_RANKS` here and take also the maximum rank into account (can be limited by an addon)